libraries/chain/hotstuff/finalizer.cpp

```diff
@@ -0,0 +1,259 @@
+#include <eosio/chain/hotstuff/finalizer.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <fc/log/logger_config.hpp>
+
+namespace eosio::chain {
+
+// ----------------------------------------------------------------------------------------
+block_state_ptr get_block_by_num(const fork_database_if_t::branch_type& branch, std::optional<uint32_t> block_num) {
+   if (!block_num || branch.empty())
+      return block_state_ptr{};
+
+   // a branch always contains consecutive block numbers, starting with the highest
+   uint32_t first = branch[0]->block_num();
+   uint32_t dist  = first - *block_num;
+   return dist < branch.size() ? branch[dist] : block_state_ptr{};
+}
+
+// ----------------------------------------------------------------------------------------
+bool extends(const fork_database_if_t& fork_db, const block_state_ptr& descendant, const block_id_type& ancestor) {
+   if (ancestor.empty())
+      return false;
+   auto cur = fork_db.get_block_header(descendant->previous()); // use `get_block_header` so we can get the root
+   while (cur) {
+      if (cur->id() == ancestor)
+         return true;
+      cur = fork_db.get_block_header(cur->previous());
+   }
+   return false;
+}
+
+// ----------------------------------------------------------------------------------------
+finalizer::vote_decision finalizer::decide_vote(const block_state_ptr& proposal, const fork_database_if_t& fork_db) {
+   bool safety_check   = false;
+   bool liveness_check = false;
+
+   auto p_branch = fork_db.fetch_branch(proposal->id());
```

Comment:
No reason to grab the branch unless `monotony_check && (liveness_check || safety_check)`, this can be moved down into the `if`.